crates/engine/tree/src/tree/mod.rs

```diff
@@ -2680,10 +2681,45 @@ where
     ) -> Result<TrieInput, ParallelStateRootError> {
         let mut input = TrieInput::default();
 
-        if let Some((historical, blocks)) = self.state.tree_state.blocks_by_hash(parent_hash) {
+        let provider = consistent_view.provider_ro()?;
+        let best_block_number = provider.best_block_number()?;
+
+        let in_memory_blocks = self
+            .state
+            .tree_state
+            .blocks_by_hash(parent_hash)
+            .map(|(_, blocks)| {
+                let mut filtered_blocks = Vec::with_capacity(blocks.len());
+                let mut historical = None;
+                for block in blocks {
+                    let recovered_block = block.recovered_block();
+
+                    // Take only those blocks that are either:
+                    // 1. Higher than the highest database block
+                    // 2. Or have no associated database block, meaning that it's on a non-canonical
+                    //    chain
+                    if recovered_block.number() > best_block_number ||
+                        Some(recovered_block.hash()) !=
+                            provider.block_hash(recovered_block.number())?
+                    {
+                        historical = Some(recovered_block.parent_hash());
+                        filtered_blocks.push(block);
+                    } else {
+                        // Since we're iterating from highest to lowest, we can stop here, meaning
+                        // that we started seeing blocks that are present in the database.
+                        break
+                    }
+                }
+
```

Comment:
I was thinking instead of doing this we maybe just check the first value in the vec returned by `blocks_by_hash`, we `pop_front` if the block is in the db, stop if we find one that is not in the db